internal/jennies/java/deserializers.go

```diff
@@ -0,0 +1,164 @@
+package java
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/grafana/codejen"
+	"github.com/grafana/cog/internal/ast"
+	"github.com/grafana/cog/internal/languages"
+	"github.com/grafana/cog/internal/tools"
+)
+
+type Deserializers struct {
+	config  Config
+	imports []string
+}
+
+func (jenny *Deserializers) JennyName() string {
+	return "JavaDeserializers"
+}
+
+func (jenny *Deserializers) Generate(context languages.Context) (codejen.Files, error) {
+	deserialisers := make(codejen.Files, 0)
+	for _, schema := range context.Schemas {
+		var hasErr error
+		schema.Objects.Iterate(func(key string, obj ast.Object) {
+			if objectNeedsCustomDeserialiser(context, obj) {
+				f, err := jenny.genCustomDeserialiser(context, obj)
+				if err != nil {
+					hasErr = err
+				} else {
+					deserialisers = append(deserialisers, *f)
+				}
+			}
+		})
+		if hasErr != nil {
+			return nil, hasErr
+		}
+	}
+
+	return deserialisers, nil
+}
+
+func (jenny *Deserializers) genCustomDeserialiser(context languages.Context, obj ast.Object) (*codejen.File, error) {
+	if obj.Type.IsStruct() && obj.Type.HasHint(ast.HintDisjunctionOfScalars) {
+		return jenny.genDisjunctionsDeserialiser(obj, "disjunctions_of_scalars")
+	}
+
+	if obj.Type.IsStruct() && obj.Type.HasHint(ast.HintDiscriminatedDisjunctionOfRefs) {
+		return jenny.genDisjunctionsDeserialiser(obj, "disjunctions_of_refs")
+	}
+
+	return jenny.genDataqueryDeserialiser(context, obj)
+}
+
+func (jenny *Deserializers) genDataqueryDeserialiser(context languages.Context, obj ast.Object) (*codejen.File, error) {
+	buf := bytes.Buffer{}
+
+	jenny.imports = jenny.genImports(obj)
+	dataqueryCode, err := jenny.genDataqueryCode(context, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := templates.ExecuteTemplate(&buf, "marshalling/unmarshalling.tmpl", Unmarshalling{
+		Package:                   jenny.formatPackage(obj.SelfRef.ReferredPkg),
+		Name:                      obj.Name,
+		ShouldUnmarshallingPanels: obj.SelfRef.ReferredPkg == "dashboard" && obj.Name == "Panel",
+		Imports:                   jenny.imports,
+		Fields:                    obj.Type.AsStruct().Fields,
+		DataqueryUnmarshalling:    dataqueryCode,
+	}); err != nil {
+		return nil, fmt.Errorf("failed executing template: %w", err)
+	}
+
+	path := filepath.Join(jenny.config.ProjectPath, obj.SelfRef.ReferredPkg, fmt.Sprintf("%sDeserializer.java", obj.SelfRef.ReferredType))
+	return codejen.NewFile(path, buf.Bytes(), jenny), nil
+}
+
+func (jenny *Deserializers) genDataqueryCode(context languages.Context, obj ast.Object) (DataqueryUnmarshalling, error) {
+	for _, field := range obj.Type.AsStruct().Fields {
+		composableSlotType, resolved := context.ResolveToComposableSlot(field.Type)
+		if !resolved {
+			continue
+		}
+
+		if composableSlotType.AsComposableSlot().Variant == ast.SchemaVariantDataQuery {
+			return jenny.renderUnmarshalDataqueryField(obj, field), nil
```

Comment:
return or continue? Is there only one composable slot per struct?

Reply:
It should be a continue, right. It was working because we don't have more than one case right now 😅